### PR TITLE
Bunch of improvments

### PR DESCRIPTION
**Pull request by MaxGraey**
- speedup blake2b64 (I hope so, didn't really benchmarking)
- update AssemblyScript to 14.0.0
- other improvments and refactorings

### CODE REVIEW

File: src/blake2b/blake2b64.ts [hunk 1]
```diff
@@ -153,7 +223,7 @@ export function blake2bInit(outlen: u32, key: Uint8Array | null): Context {
 // Updates a BLAKE2b streaming hash
 // Requires hash context and Uint8Array (byte array)
 export function blake2bUpdate(ctx: Context, input: Uint8Array): void {
-  for (let i = 0; i < input.length; i++) {
+  for (let i = 0, len = input.length; i < len; i++) {
```
Comment (gzuidhof) [inline src/blake2b/blake2b64.ts:224]: Does something like this actually matter? I would expect the compiler to pick this up, or is it not that smart?

File: src/blake2b/blake2b64.ts [hunk 2]
```diff
@@ -201,22 +274,28 @@ export function blake2b(input: Uint8Array, key: Uint8Array | null = null, outlen
...
   // }
-  // ctx.b = input; 
+  // ctx.b = input;
   // We don't actually have to reset these as they are overwritten and we never stream..
   // ctx.b.set(input);
   // ctx.m.fill(0);
   // ctx.v.fill(0);
 
-  ctx.h[0] ^= 0x01010000 ^ (ctx.outlen as u64);
+  h[0] = BLAKE2B_IV_0 ^ 0x01010000 ^ (ctx.outlen as u64);
```
Comment (gzuidhof) [inline src/blake2b/blake2b64.ts]: Nit: You could move this up to where  ```// unchecked(h[0] = BLAKE2B_IV_0);``` currently is (line 281), and make it unchecked

**Comment by gzuidhof** [comment]
Wow, thank you so much! This is really amazing.

I didn't expect there to be so much improvement in speed left without making it unreadable, but you've done a really good job. I am super impressed because the code you optimized wasn't the most clear anymore after I iterated on it a lot for speed (sorry about that!).

My poor man's benchmark has been to edit the expectation [here](https://github.com/gzuidhof/friendly-pow/blob/master/test/solver.astest.ts#L24) and make it really high (e.g. 10 million). And then run:
```
npm run test:as -- --file solver.astest.ts
```

This runs in Node, I've found that it's up to in Chrome or Firefox it can be another 50% faster (my node version is 12.13.1). 

## Benchmarks

### Without your changes (3 runs):
```
[Actual]: 2242152.466367713
[Actual]: 2309468.8221709006
[Actual]: 2227171.4922049
```
### With your changes (3 runs)
```
[Actual]: 4032258.064516129
[Actual]: 4016064.2570281127
[Actual]: 3952569.169960474
```

We went from 2.25 million hashes to 4 million hashes per second, really impressive! I didn't expect this to be possible without SIMD. The build size went from 3.44KB to 3.24KB, so even there it's a small improvement 🚀.

**Review by gzuidhof — APPROVED** [review]
Thank you again :), I only left a question and a small nit, really keen on merging this and updating the CAPTCHA widget upstream and see how much faster it is in a browser. I learned a lot by going through the changes.

File: src/blake2b/blake2b64.ts (hunk 1 above)
Comment (MaxGraey) [reply src/blake2b/blake2b64.ts:224]: Yes, it's little bit matter. AS don't do such high-level CSE transform. So better cache `len` manually for now. But it' really small improvement

**Comment by MaxGraey** [comment]
What do you think btw made external assemblyscript module for `blake2b64` (as-blake2b)? Similar to [`as-sha256`](https://github.com/ChainSafe/as-sha256)

**Comment by gzuidhof** [comment]
Yeah I'd be happy to make it an external module, but we do need to verify that the implementation is actually correct.

All my tests and use cases are passing in a single message that is under 128 bytes, who knows whether it is also correct for longer messages, and messages that are streamed.

**Comment by gzuidhof** [comment]
Feel free to merge when you are ready! thanks again

**Comment by MaxGraey** [comment]
> All my tests and use cases are passing in a single message that is under 128 bytes

Oh, definitely need more tests)

> Feel free to merge when you are ready

It's ready

**Pull request merged**